ballot/src/lib.rs

```diff
@@ -0,0 +1,173 @@
+use raiko_lib::proof_type::ProofType;
+use serde::{Deserialize, Serialize};
+use std::collections::BTreeMap;
+
+/// Ballot is a proof type selection mechanism using the block number as the random number.
+///
+/// Note that Ballot is deterministic and does not require any randomness, it uses the deterministic
+/// block number to select the proof type, so that we can replay the proof type selection for the same
+/// block number and get the same result.
+#[derive(Debug, Clone, Default, Serialize, Deserialize)]
+pub struct Ballot {
+    /// The proof type to always use
+    always_proof_type: ProofType,
+
+    /// A map of ProofType to their probabilities (between 0 and 1)
+    probabilities: BTreeMap<ProofType, f64>,
+}
+
+impl Ballot {
+    /// Create a new Ballot
+    pub fn new(
+        always_proof_type: ProofType,
+        probs: BTreeMap<ProofType, f64>,
+    ) -> Result<Self, String> {
+        let ballot = Self {
+            always_proof_type,
+            probabilities: probs,
+        };
+        ballot.validate()?;
+        Ok(ballot)
+    }
+
+    pub fn validate(&self) -> Result<(), String> {
+        // Validate the allowed proof types
+        let allowed_proof_types = vec![ProofType::Sp1, ProofType::Risc0];
+        for (proof_type, _) in self.probabilities.iter() {
+            if !allowed_proof_types.contains(proof_type) {
+                return Err(format!("Not allowed prob proof type: {proof_type}"));
+            }
+        }
+
+        // Validate the always proof type
+        if !allowed_proof_types.contains(&self.always_proof_type) {
+            return Err(format!(
+                "Not allowed always proof type: {}",
+                self.always_proof_type
+            ));
+        }
+
+        // Validate each probability
+        for (&proof_type, &prob) in self.probabilities.iter() {
+            if !(0.0..=1.0).contains(&prob) {
+                return Err(format!(
+                    "Invalid probability value {} for proof type {:?}, must be between 0 and 1",
+                    prob, proof_type
+                ));
+            }
+        }
+
+        // Validate the total probability
+        let total_prob: f64 = self.probabilities.values().sum();
+        if total_prob > 1.0 {
+            return Err(format!(
+                "Total probability must be less than or equal to 1.0, but got {total_prob}"
+            ));
+        }
+
+        Ok(())
+    }
+
+    /// Draw proof types based on the block number, returning the always proof type and the proof type
+    /// selected by the block number.
+    pub fn draw(&self, block_number: u64) -> (ProofType, Option<ProofType>) {
+        const PRECISION: u64 = 10_000;
+        let remainder = block_number % PRECISION;
+
+        let mut cumulative_prob = 0.0;
+        for (proof_type, &prob) in self.probabilities.iter() {
+            cumulative_prob += prob;
+            if remainder < (cumulative_prob * PRECISION as f64).round() as u64 {
```

Comment:
Assuming sp1 is 5% and r0 is also 5%, if you use 10000 as the precision value (PRECISION), the first 500 blocks will be assigned to sp1, the next 500 to risc0, and the remaining 9000 blocks will remain idle. Although the ratio is still 10%, this large-scale burst behavior doesn't seem ideal (for now), as managing 500 remote ZK tasks can be error-prone.
Especially in product env, I recommend adopting a more time-distributed draw instead of this burst allocation. However, before we find a better solution, using 100 as the precision value seems also acceptable, as handling a burst of 5/5 tasks and idol for 90 is much easier

Reply:
Thanks for your pointing out. Fixed.